archive/tar_stream.ts

```diff
@@ -0,0 +1,466 @@
+}
+
+/**
+ * A union type merging all the TarStream interfaces that can be piped into the
+ * TarStream class.
+ */
+export type TarStreamInput = TarStreamFile | TarStreamDir;
+type TarStreamInputInternal =
+  & (Omit<TarStreamFile, "pathname"> | Omit<TarStreamDir, "pathname">)
+  & { pathname: [Uint8Array, Uint8Array] };
+
+/**
+ * The options that can go along with a file or directory.
+ */
+export interface TarStreamOptions {
+  /**
+   * An octal literal.
+   * Defaults to 755 for directories and 644 for files.
+   */
+  mode?: number;
+  /**
+   * An octal literal.
+   * @default {0}
+   */
+  uid?: number;
+  /**
+   * An octal literal.
+   * @default {0}
+   */
+  gid?: number;
+  /**
+   * A number of seconds since the start of epoch. Avoid negative values.
+   * Defaults to the current time in seconds.
+   */
+  mtime?: number;
+  /**
+   * An ASCII string. Should be used in preference of uid.
+   * @default {''}
+   */
+  uname?: string;
+  /**
+   * An ASCII string. Should be used in preference of gid.
+   * @default {''}
+   */
+  gname?: string;
+  /**
+   * The major number for character device.
+   * @default {''}
+   */
+  devmajor?: string;
+  /**
+   * The minor number for block device entry.
+   * @default {''}
+   */
+  devminor?: string;
+}
+
+const SLASH_CODE_POINT = "/".charCodeAt(0);
+
+/**
+ * ### Overview
+ * A TransformStream to create a tar archive.  Tar archives allow for storing
+ * multiple files in a single file (called an archive, or sometimes a tarball).
+ *   These archives typically have a single '.tar' extension.  This
+ * implementation follows the [FreeBSD 15.0](https://man.freebsd.org/cgi/man.cgi?query=tar&sektion=5&apropos=0&manpath=FreeBSD+15.0-CURRENT) spec.
+ *
+ * ### File Format & Limitations
+ * The ustar file format is used for creating the tar archive.  While this
+ * format is compatible with most tar readers, the format has several
+ * limitations, including:
+ * - Pathnames must be at most 256 characters.
+ * - Files must be at most 8 GiBs in size, or 64 GiBs if `sizeExtension` is set
+ * to true.
+ * - Sparse files are not supported.
+ *
+ * ### Usage
+ * TarStream may throw an error for several reasons. A few of those are:
+ * - The pathname is invalid.
+ * - The size provided does not match that of the iterable's length.
+ *
+ * ### Compression
+ * Tar archives are not compressed by default.  If you'd like to compress the
+ * archive, you may do so by piping it through a compression stream.
+ *
+ * @example Usage
+ * ```ts no-eval
+ * import { TarStream } from "@std/archive/tar-stream";
+ *
+ * await ReadableStream.from([
+ *   {
+ *     pathname: 'potato/'
+ *   },
+ *   {
+ *     pathname: 'deno.json',
+ *     size: (await Deno.stat('deno.json')).size,
+ *     iterable: (await Deno.open('deno.json')).readable
+ *   },
+ *   {
+ *     pathname: 'deno.lock',
+ *     size: (await Deno.stat('deno.lock')).size,
+ *     iterable: (await Deno.open('deno.lock')).readable
+ *   }
+ * ])
```

Comment:
Including a file nested in a subdirectory would help clarify the format of the `pathname` string.